src/main/flight/wind_estimator.c

```diff
@@ -0,0 +1,155 @@
+
+#include "fc/config.h"
+#include "fc/runtime_config.h"
+
+#include "flight/imu.h"
+
+#include "io/gps.h"
+
+// Based on WindEstimation.pdf paper
+
+static bool hasValidWindEstimate = false;
+static float estimatedWind[XYZ_AXIS_COUNT] = {0, 0, 0};    // wind velocity vectors in cm / sec in earth frame
+static float lastGroundVelocity[XYZ_AXIS_COUNT];
+static float lastFuselageDirection[XYZ_AXIS_COUNT];
+
+bool isEstimatedWindSpeedValid(void)
+{
+    // TODO: Add a timeout. Estimated wind should expire if
+    // if we can't update it for an extended time.
+    return hasValidWindEstimate;
+}
+
+float getEstimatedWindSpeed(int axis)
+{
+    return estimatedWind[axis];
+}
+
+void updateWindEstimator(timeUs_t currentTimeUs)
+{
+    static timeUs_t lastUpdateUs = 0;
+
+    if (!STATE(FIXED_WING) ||
+        !isGPSHeadingValid() ||
+        !gpsSol.flags.validVelNE ||
+        !gpsSol.flags.validVelD) {
+        return;
+    }
+
+    float groundVelocity[XYZ_AXIS_COUNT];
+    float groundVelocityDiff[XYZ_AXIS_COUNT];
+    float groundVelocitySum[XYZ_AXIS_COUNT];
+
+    float fuselageDirection[XYZ_AXIS_COUNT];
+    float fuselageDirectionDiff[XYZ_AXIS_COUNT];
+    float fuselageDirectionSum[XYZ_AXIS_COUNT];
+
+    // Get current 3D velocity from GPS in cm/s
+    // relative to earth frame
+    groundVelocity[X] = gpsSol.velNED[X];
+    groundVelocity[Y] = gpsSol.velNED[Y];
+    groundVelocity[Z] = gpsSol.velNED[Z];
+
+    // Fuselage direction in earth frame
+    fuselageDirection[X] = rMat[0][0];
+    fuselageDirection[Y] = rMat[1][0];
+    fuselageDirection[Z] = rMat[2][0];
+
+    timeDelta_t timeDelta = cmpTimeUs(currentTimeUs, lastUpdateUs);
+    // scrap our data and start over if we're taking too long to get a direction change
+    if (lastUpdateUs == 0 || timeDelta > 2 * USECS_PER_SEC) {
+
+        lastUpdateUs = currentTimeUs;
+
+        memcpy(lastFuselageDirection, fuselageDirection, sizeof(lastFuselageDirection));
+        memcpy(lastGroundVelocity, groundVelocity, sizeof(lastGroundVelocity));
+        return;
+    }
+
+    fuselageDirectionDiff[X] = fuselageDirection[X] - lastFuselageDirection[X];
+    fuselageDirectionDiff[Y] = fuselageDirection[Y] - lastFuselageDirection[Y];
+    fuselageDirectionDiff[Z] = fuselageDirection[Z] - lastFuselageDirection[Z];
+
+    float diff_length = sqrtf(sq(fuselageDirectionDiff[X]) + sq(fuselageDirectionDiff[Y]) + sq(fuselageDirectionDiff[Z]));
+    // Very small changes in attitude will result in a denominator
+    // very close to zero which will introduce too much error in the
+    // estimation.
+    //
+    // TODO: Is 0.2f an adequate threshold?
+    if (diff_length > 0.2f) {
+        // when turning, use the attitude response to estimate wind speed
+        groundVelocityDiff[X] = groundVelocity[X] - lastGroundVelocity[X];
+        groundVelocityDiff[Y] = groundVelocity[Y] - lastGroundVelocity[Y];
+        groundVelocityDiff[Z] = groundVelocity[X] - lastGroundVelocity[Z];
```

Comment:
Do we really need to calculate Z axis here? I'm not sure that wind can blow up or down.

https://github.com/MatrixPilot/MatrixPilot/blob/9e3f52ecd7ae09e24c4eddaec19625f77a0f9401/libDCM/estWind.c line 85 they don't calculate it.